addons/hr/models/hr_employee.py

```diff
@@ -535,6 +535,43 @@ def _get_tz(self):
                self.company_id.resource_calendar_id.tz or\
                'UTC'
 
+    def _get_calendar_tz(self, date_from=None):
+        # same as ``_get_tz`` but prioritze the date on the employee's calendar
+        self.ensure_one()
+        if date_from:  # contract overload has different signature
+            calendar = self._get_calendar(date_from)
+        else:
+            calendar = self._get_calendar()
+        return calendar.tz or\
+               self.tz
+
+    @api.model
+    def _get_timezones_query(self):
```

Comment:
I don't like the fact that we have parallel methods to retrieve the timezone using the ORM or pure SQL. For me everything should be done using the ORM, for maintainability purposes. If performances are bad at some points, this is probably either because the code is badly written (search in loops, etc) or because some indexes are missing. 

Pure SQL is also bypassing access rules, so you could retrieve attendance information from attendances that are unavailable for you (other company, etc) or for records that are not supposed to be taken into account (archived, etc).